src/Compilers/CSharp/Portable/Binder/BindingDiagnosticBag.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class BindingDiagnosticBagExtensions
+    {
+        internal static CSDiagnosticInfo Add(this CodeAnalysis.BindingDiagnosticBag diagnostics, ErrorCode code, Location location)
+        {
+            var info = new CSDiagnosticInfo(code);
+            diagnostics.DiagnosticBag?.Add(new CSDiagnostic(info, location));
+            return info;
+        }
+
+        internal static CSDiagnosticInfo Add(this CodeAnalysis.BindingDiagnosticBag diagnostics, ErrorCode code, Location location, params object[] args)
+        {
+            var info = new CSDiagnosticInfo(code, args);
+            diagnostics.DiagnosticBag?.Add(new CSDiagnostic(info, location));
+            return info;
+        }
+
+        internal static CSDiagnosticInfo Add(this CodeAnalysis.BindingDiagnosticBag diagnostics, ErrorCode code, Location location, ImmutableArray<Symbol> symbols, params object[] args)
+        {
+            var info = new CSDiagnosticInfo(code, args, symbols, ImmutableArray<global::Microsoft.CodeAnalysis.Location>.Empty);
+            diagnostics.DiagnosticBag?.Add(new CSDiagnostic(info, location));
+            return info;
+        }
+
+        internal static void Add(this CodeAnalysis.BindingDiagnosticBag diagnostics, DiagnosticInfo? info, Location location)
+        {
+            if (info is object)
+            {
+                diagnostics.DiagnosticBag?.Add(info, location);
+            }
+        }
+
+        internal static void AddForSymbol(this ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo, Symbol? symbol, bool addDiagnostics = true, bool addDependencies = true)
+        {
+            if (symbol is null || useSiteInfo.IsDiscarded || !(addDiagnostics || addDependencies))
+            {
+                return;
+            }
+
+            var info = symbol.GetUseSiteInfo();
+
+            if (addDiagnostics)
+            {
+                useSiteInfo.AddDiagnostics(info);
+            }
+            else
+            {
+                Debug.Assert(info.DiagnosticInfo == null || addDependencies); // It would be strange to drop diagnostics, but record dependencies
```

Comment:
>addDependencies [](start = 60, length = 15)

If `addDiagnostics` is false, we've already guaranteed that `addDependencies` is true in the `if` above. This assertion will never fail.

Reply:
> If addDiagnostics is false, we've already guaranteed that addDependencies is true in the if above. This assertion will never fail.

Thanks, I'll remove the assert.

---
In reply to: [361554554](https://github.com/dotnet/roslyn/pull/40560#discussion_r361554554) [](ancestors = 361554554)

Reply:
The Assert is adjusted in [https://github.com/dotnet/roslyn/pull/40640](https://github.com/dotnet/roslyn/pull/40640)

---
In reply to: [361557360](https://github.com/dotnet/roslyn/pull/40560#discussion_r361557360) [](ancestors = 361557360,361554554)